### ISSUE TITLE
Destroy app often requires marathon service restart

### ISSUE THREAD

**Issue by doctapp** [mesosphere/marathon#1317]
We've used Marathon for a while now, but since 0.8.0 and 0.8.1 came out, we have experienced the following bug with both versions.

After having manually started (PUT) and destroyed (DELETE) an app multiple times (5-10), Marathon reports the app as started, the PUT succeeded, while the app doesn't show up in the Marathon Web UI nor gets scheduled in Mesos (by looking at its web ui). Note that starting an app with a different id works fine.

This behavior is very consistent for an app that is in development mode which you restart with a different code base fast say within 1min of shutting the previous instance down.

The only solution is to manually restart the Marathon service. Sometimes, we even need to clean up the zookeeper state using zookeeper-client.

What can we do to prevent this?

Thanks

**Comment by kolloch** [comment]
Hi @doctapp,

thanks for reporting this. Unfortunately, I could not reproduce that so far on the current master.

I used the following app definition:

```
{
  "id" : "app",
  "cmd" : "python -m SimpleHTTPServer $PORT",
  "ports" : [ 0 ],
  "cpus" : 0.1,
  "mem" : 32,
  "instances" : 1,
  "healthChecks" : [ {
    "protocol" : "HTTP",
    "portIndex" : 0,
    "path" : "/",
    "gracePeriodSeconds" : 5,
    "intervalSeconds" : 5,
    "maxConsecutiveFailures" : 3
  } ]
}
```

I repeated the following the requests 50x sequentially fast after each other, irrespective of the resulting status code:

```
PUT http://127.0.0.1:8080/v2/apps/app
DELETE http://127.0.0.1:8080/v2/apps/app
```

I tried the same with waiting fall all deployments finishing in between.

After every test run, I was able to destroy all ups with the API and creating the app anew. I didn't have to restart marathon.

Do my attempts to reproduce the problem look reasonable or would you suggest something else?

Can you send us your log files for the time in which the error occurred? Maybe together with stack traces of a broken marathon (jstack or kill -3)?

**Comment by doctapp** [comment]
I can consistently reproduce this with your example. Here's the script I'm running:
---deploy.sh---
#!/usr/bin/env bash
DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
NAME=test-marathon
MARATHON=http://myserver:8080

pushd $DIR/.. > /dev/null

case $1 in
        start)
            JSON=$(cat << EOF
            {
              "id" : "$NAME",
              "cmd" : "python -m SimpleHTTPServer $PORT",
              "ports" : [ 0 ],
              "cpus" : 0.1,
              "mem" : 32,
              "instances" : 1,
              "healthChecks" : [ {
                "protocol" : "HTTP",
                "portIndex" : 0,
                "path" : "/",
                "gracePeriodSeconds" : 5,
                "intervalSeconds" : 5,
                "maxConsecutiveFailures" : 3
              } ]
            }
EOF
)
            echo "$JSON" | http "$MARATHON/v2/apps" 
            ;;

        stop)
            http DELETE $MARATHON/v2/apps/$NAME
            ;;

        *)
            echo $"Usage: $0 {start|stop}"
            exit 1
esac

popd > /dev/null
---deploy.sh---

Here's the script execution history:
---command.log---
~/dev$ ./deploy.sh start
HTTP/1.1 201 Created
Cache-Control: no-cache, no-store, must-revalidate
Content-Type: application/json
Expires: 0
Location: http://myserver:8080/v2/apps/test-marathon
Pragma: no-cache
Server: Jetty(8.y.z-SNAPSHOT)
Transfer-Encoding: chunked

{
    "args": null, 
    "backoffFactor": 1.15, 
    "backoffSeconds": 1, 
    "cmd": "python -m SimpleHTTPServer ", 
    "constraints": [], 
    "container": null, 
    "cpus": 0.1, 
    "dependencies": [], 
    "deployments": [
        {
            "id": "84623c6d-8f6d-40bc-afca-e6ca3df22991"
        }
    ], 
    "disk": 0.0, 
    "env": {}, 
    "executor": "", 
    "healthChecks": [
        {
            "command": null, 
            "gracePeriodSeconds": 5, 
            "intervalSeconds": 5, 
            "maxConsecutiveFailures": 3, 
            "path": "/", 
            "portIndex": 0, 
            "protocol": "HTTP", 
            "timeoutSeconds": 20
        }
    ], 
    "id": "/test-marathon", 
    "instances": 1, 
    "labels": {}, 
    "maxLaunchDelaySeconds": 3600, 
    "mem": 32.0, 
    "ports": [
        0
    ], 
    "requirePorts": false, 
    "storeUrls": [], 
    "tasks": [], 
    "tasksHealthy": 0, 
    "tasksRunning": 0, 
    "tasksStaged": 0, 
    "tasksUnhealthy": 0, 
    "upgradeStrategy": {
        "maximumOverCapacity": 1.0, 
        "minimumHealthCapacity": 1.0
    }, 
    "uris": [], 
    "user": null, 
    "version": "2015-03-20T18:28:03.924Z"
}

~/dev$ ./deploy.sh stop
HTTP/1.1 200 OK
Cache-Control: no-cache, no-store, must-revalidate
Content-Type: application/json
Expires: 0
Pragma: no-cache
Server: Jetty(8.y.z-SNAPSHOT)
Transfer-Encoding: chunked

{
    "deploymentId": "782c80ad-a24f-4787-bf07-a9edc0d3699d", 
    "version": "2015-03-20T18:28:21.689Z"
}

~/dev$ ./deploy.sh start
HTTP/1.1 201 Created
Cache-Control: no-cache, no-store, must-revalidate
Content-Type: application/json
Expires: 0
Location: http://myserver:8080/v2/apps/test-marathon
Pragma: no-cache
Server: Jetty(8.y.z-SNAPSHOT)
Transfer-Encoding: chunked

{
    "args": null, 
    "backoffFactor": 1.15, 
    "backoffSeconds": 1, 
    "cmd": "python -m SimpleHTTPServer ", 
    "constraints": [], 
    "container": null, 
    "cpus": 0.1, 
    "dependencies": [], 
    "deployments": [
        {
            "id": "2a21b007-dabe-4c2a-b3c9-66ad02bd7301"
        }
    ], 
    "disk": 0.0, 
    "env": {}, 
    "executor": "", 
    "healthChecks": [
        {
            "command": null, 
            "gracePeriodSeconds": 5, 
            "intervalSeconds": 5, 
            "maxConsecutiveFailures": 3, 
            "path": "/", 
            "portIndex": 0, 
            "protocol": "HTTP", 
            "timeoutSeconds": 20
        }
    ], 
    "id": "/test-marathon", 
    "instances": 1, 
    "labels": {}, 
    "maxLaunchDelaySeconds": 3600, 
    "mem": 32.0, 
    "ports": [
        0
    ], 
    "requirePorts": false, 
    "storeUrls": [], 
    "tasks": [], 
    "tasksHealthy": 0, 
    "tasksRunning": 0, 
    "tasksStaged": 0, 
    "tasksUnhealthy": 0, 
    "upgradeStrategy": {
        "maximumOverCapacity": 1.0, 
        "minimumHealthCapacity": 1.0
    }, 
    "uris": [], 
    "user": null, 
    "version": "2015-03-20T18:28:33.207Z"
}

~/dev$ ./deploy.sh stop
HTTP/1.1 200 OK
Cache-Control: no-cache, no-store, must-revalidate
Content-Type: application/json
Expires: 0
Pragma: no-cache
Server: Jetty(8.y.z-SNAPSHOT)
Transfer-Encoding: chunked

{
    "deploymentId": "1bf57b60-5b05-48a3-b9bc-3695fa45679a", 
    "version": "2015-03-20T18:28:42.471Z"
}

~/dev$ ./deploy.sh start
HTTP/1.1 504 
Cache-Control: no-cache, no-store, must-revalidate
Content-Type: application/json
Expires: 0
Pragma: no-cache
Server: Jetty(8.y.z-SNAPSHOT)
Transfer-Encoding: chunked

{
    "message": "Failed to wait for future within timeout"
}

~/dev$ ./deploy.sh start
HTTP/1.1 201 Created
Cache-Control: no-cache, no-store, must-revalidate
Content-Type: application/json
Expires: 0
Location: http://myserver:8080/v2/apps/test-marathon
Pragma: no-cache
Server: Jetty(8.y.z-SNAPSHOT)
Transfer-Encoding: chunked

{
    "args": null, 
    "backoffFactor": 1.15, 
    "backoffSeconds": 1, 
    "cmd": "python -m SimpleHTTPServer ", 
    "constraints": [], 
    "container": null, 
    "cpus": 0.1, 
    "dependencies": [], 
    "deployments": [], 
    "disk": 0.0, 
    "env": {}, 
    "executor": "", 
    "healthChecks": [
        {
            "command": null, 
            "gracePeriodSeconds": 5, 
            "intervalSeconds": 5, 
            "maxConsecutiveFailures": 3, 
            "path": "/", 
            "portIndex": 0, 
            "protocol": "HTTP", 
            "timeoutSeconds": 20
        }
    ], 
    "id": "/test-marathon", 
    "instances": 1, 
    "labels": {}, 
    "maxLaunchDelaySeconds": 3600, 
    "mem": 32.0, 
    "ports": [
        0
    ], 
    "requirePorts": false, 
    "storeUrls": [], 
    "tasks": [], 
    "tasksHealthy": 0, 
    "tasksRunning": 0, 
    "tasksStaged": 0, 
    "tasksUnhealthy": 0, 
    "upgradeStrategy": {
        "maximumOverCapacity": 1.0, 
        "minimumHealthCapacity": 1.0
    }, 
    "uris": [], 
    "user": null, 
    "version": "2015-03-20T18:29:04.090Z"
}

~/dev$ ./deploy.sh stop
HTTP/1.1 404 Not Found
Cache-Control: no-cache, no-store, must-revalidate
Content-Type: application/json
Expires: 0
Pragma: no-cache
Server: Jetty(8.y.z-SNAPSHOT)
Transfer-Encoding: chunked

{
    "message": "App '/test-marathon' does not exist"
}

~/dev$ ./deploy.sh start
HTTP/1.1 201 Created
Cache-Control: no-cache, no-store, must-revalidate
Content-Type: application/json
Expires: 0
Location: http://myserver:8080/v2/apps/test-marathon
Pragma: no-cache
Server: Jetty(8.y.z-SNAPSHOT)
Transfer-Encoding: chunked

{
    "args": null, 
    "backoffFactor": 1.15, 
    "backoffSeconds": 1, 
    "cmd": "python -m SimpleHTTPServer ", 
    "constraints": [], 
    "container": null, 
    "cpus": 0.1, 
    "dependencies": [], 
    "deployments": [], 
    "disk": 0.0, 
    "env": {}, 
    "executor": "", 
    "healthChecks": [
        {
            "command": null, 
            "gracePeriodSeconds": 5, 
            "intervalSeconds": 5, 
            "maxConsecutiveFailures": 3, 
            "path": "/", 
            "portIndex": 0, 
            "protocol": "HTTP", 
            "timeoutSeconds": 20
        }
    ], 
    "id": "/test-marathon", 
    "instances": 1, 
    "labels": {}, 
    "maxLaunchDelaySeconds": 3600, 
    "mem": 32.0, 
    "ports": [
        0
    ], 
    "requirePorts": false, 
    "storeUrls": [], 
    "tasks": [], 
    "tasksHealthy": 0, 
    "tasksRunning": 0, 
    "tasksStaged": 0, 
    "tasksUnhealthy": 0, 
    "upgradeStrategy": {
        "maximumOverCapacity": 1.0, 
        "minimumHealthCapacity": 1.0
    }, 
    "uris": [], 
    "user": null, 
    "version": "2015-03-20T18:29:34.195Z"
}
---command.log---

As soon as I hit the "message": "Failed to wait for future within timeout", Marathon accepts the job (same output as regular run), but does not run them nor show them in it's UI. Deleting the job always fails with the "message": "App '/test-marathon' does not exist".

I've looked at the /var/log/upstart/marathon.log, but it only contains the following:
run_jar --zk zk://myserver:2181/marathon --master zk://myserver:2181/mesos
run_jar --zk zk://myserver:2181/marathon --master zk://myserver:2181/mesos

Do you need more info?

**Comment by kolloch** [comment]
Hi @doctapp,

thanks for your investigation&report. That's really great.

At first sight, that looks pretty similar to what I did, though. I will look at it again on Monday.

I only tried to reproduce it on master so far. I'll check with 0.81. If you want, you can also check it with the master version.

Marathon logs a lot more. The location depends on your setup. Maybe grep /var/log for marathon.

**Comment by kolloch** [comment]
Hi @doctapp,

I tried to reproduce your problem with my local development setup. Do you have a complicated setup? I changed your script slightly by quoting the `$PORT` variable reference (otherwise the dynamically allocated port is not used and the health check fails).

I used Marathon version 0.8.1. Unfortunately, I could not reproduce your error.

Maybe you can avoid the problem by using another request pattern. If your app definition changes between the restarts, you can just use a `PUT` request for the new app definition and the deployment will start.

If your app definition does not change but the version changes implicitly (because the new software version is fetched or startup or something like that), you can use a [POST /v2/apps/{appId}/restart](https://mesosphere.github.io/marathon/docs/rest-api.html#post-/v2/apps/%7Bappid%7D/restart) request.

**Comment by doctapp** [comment]
I retried this morning and got this:
~/dev$ ./deploy-test.sh start
HTTP/1.1 409 Conflict
Cache-Control: no-cache, no-store, must-revalidate
Content-Type: application/json
Expires: 0
Pragma: no-cache
Server: Jetty(8.y.z-SNAPSHOT)
Transfer-Encoding: chunked

{
    "deployments": [], 
    "message": "App is locked by one or more deployments. Override with the option '?force=true'. View details at '/v2/deployments/<DEPLOYMENT_ID>'."
}

Once Marathon restarted, I corrected the \$PORT (which is ok for other scripts!) and I could no longer reproduce the problem. Our apps didn't use health, so that must be the problem. Therefore, I'll use health for my apps, but this behavior should be documented or somehow prevented.

Thanks for your help!

Martin
